### ISSUE TITLE
CAs generated by BOSH do not work correctly when imported into CredHub

### ISSUE THREAD

**Issue by andy-paine** [cloudfoundry-incubator/credhub#112]
## What version of the credhub server you are using?
2.5.6 (also tested on 2.5.11)

## What version of the credhub cli you are using?
2.6.2 (also tested on 2.7.0)

NB: This was tested both times against [BUCC](https://github.com/starkandwayne/bucc)

## If you were attempting to accomplish a task, what was it you were attempting to do?
Import a CA generated by `bosh create-env` into CredHub then use that CA to issue certificates (using `credhub generate`

## What did you expect to happen?
CredHub to issue certificates with a valid cert chain back to the imported CA

## What was the actual behavior?
CredHub issues certificates that don't have a valid cert chain back to the imported CA

## Please confirm where necessary:
* [ ] I have included a log output
* [ ] My log includes an error message
* [x] I have included steps for reproduction

## Reproduction steps
(I've included the whole list of commands to run since it takes a few to demonstrate the issue)
```
# Generate a CA with BOSH and import it into CredHub
cat > credhub_import.yml.tpl <<EOF
variables:
- name: bosh_generated_ca
  type: certificate
  options:
    is_ca: true
    common_name: bosh.io

credentials:
- name: bosh_generated_ca
  type: certificate
  value: ((bosh_generated_ca))
EOF

bosh int credhub_import.yml.tpl --vars-store vars.yml > credhub_import.yml
credhub import -f credhub_import.yml
credhub generate -n bosh_issued_cert -t certificate -c "foo.com" --ca bosh_generated_ca

# Not valid certificate chain
credhub get -n bosh_issued_cert -k certificate > bosh_issued_cert.crt
credhub get -n bosh_issued_cert -k ca > bosh_generated_ca.crt
openssl verify -CAfile bosh_generated_ca.crt bosh_issued_cert.crt

# Works with openssl
credhub get -n bosh_generated_ca -k private_key > bosh_generated_ca.key
openssl req -new -keyout openssl_issued_cert.key -out openssl_issued_cert.csr -newkey rsa:2048 -nodes -subj "/CN=foo.com/"
openssl x509 -req -in openssl_issued_cert.csr -CA bosh_generated_ca.crt -CAkey bosh_generated_ca.key -CAcreateserial -out openssl_issued_cert.crt
openssl verify -CAfile bosh_generated_ca.crt openssl_issued_cert.crt

# CredHub generated cert works fine
credhub generate -n credhub_generated_ca -t certificate -c "credhub.io" --is-ca
credhub generate -n credhub_issued_cert -t certificate -c "foo.com" --ca credhub_generated_ca
credhub get -n credhub_issued_cert -k ca > credhub_generated_ca.crt
credhub get -n credhub_issued_cert -k certificate > credhub_issued_cert.crt
openssl verify -CAfile credhub_generated_ca.crt credhub_issued_cert.crt
```

We've temporarily worked around this by generating the CA with openssl and then importing that rather than generating it with BOSH and this has worked for us. We don't have the option of generating the original CA with CredHub as this is when we are running `bosh create-env` so we don't have a CredHub server yet.

**Comment by tlwr** [comment]
If you look at certificates generated by Credhub:

```
sdiff <(cat /tmp/ch.yml | bosh int --path /credentials/0/value/certificate - | openssl x509 -text) <(cat /tmp/ch.yml | bosh int --path /credentials/0/value/ca - | openssl x509 -text)
```

You'll find that the x509 authority key identifier is the same for the CA and the cert, which is correct:

```
# Cert on left, CA on right
            X509v3 Authority Key Identifier: 			            X509v3 Authority Key Identifier:
                keyid:BC:90:44:BA:14:C3:95:85:C6:AD:3D:F8:D1:	                keyid:BC:90:44:BA:14:C3:95:85:C6:AD:3D:F8:D1:
```

However if you look at certs issued by Credhub, from a cert imported from BOSH, the x509 authority key identifier is different:

```
# Cert on left, CA on right
            X509v3 Authority Key Identifier: 			            X509v3 Authority Key Identifier:
                keyid:38:0E:64:E9:4D:3C:6B:3C:DD:26:37:C1:FD: |	                keyid:58:01:C7:F3:5A:CA:17:FF:3A:3B:39:3A:4F:
```

Which causes openssl to not validate correctly (use `-issuer_checks`):

```
ssl verify -verbose -CAfile /tmp/boshissuedca.pem -purpose any -issuer_checks /tmp/boshissuedcert.pem
/tmp/boshissuedcert.pem: CN = foo.com
error 30 at 0 depth lookup:authority and subject key identifier mismatch
CN = foo.com
error 30 at 0 depth lookup:authority and subject key identifier mismatch
CN = foo.com
error 20 at 0 depth lookup:unable to get local issuer certificate
```

the slightly arcane error is loosely translated as: the subject key of the CA is not the same as the authority key of the cert.

This leads me to believe [this logic](https://github.com/cloudfoundry-incubator/credhub/blob/master/components/credentials/src/main/java/org/cloudfoundry/credhub/generators/SignedCertificateGenerator.java#L123) is not correct, although I have little experience with Credhub's codebase

**Comment by andy-paine** [comment]
That reproduces for me but specifically with the BOSH generated cert. Out of interest I generated a CA with openssl:
`openssl req -new -newkey rsa:2048 -days 365 -nodes -x509 -keyout openssl_ca.key -out openssl_ca.crt -subj "/CN=openssl CA/"`
Imported that into CredHub and the certs issued from that are fine. I had to generated the CA in a linux container though as openssl on OSX seems to generate CAs that CredHub considers invalid.

**Comment by tlwr** [comment]
See this story: https://www.pivotaltracker.com/n/projects/1977341/stories/167670716

This is allegedly fixed in 2.5.7 and later, which I need to test

**Comment by andy-paine** [comment]
Running the tests posted in my original comment against 2.6.0 produce `bosh_issued_cert.crt: OK` so it looks like this issue is fixed for my use case in that version. Thanks for your help @tlwr .

tl;dr update to >=2.6.0

**Comment by tlwr** [comment]
Would it be possible to include this change in the release notes?

**Comment by tomkennedy513** [comment]
Yes, of course, thank you for catching this
